docs/src/recipes/anti-patterns.asciidoc

```diff
@@ -0,0 +1,291 @@
+
+[gremlin-groovy,modern]
+----
+g.V().has("person","name","marko").out()
+g.V().has("person","name","marko").out("created").valueMap()
+g.V().has("software","name","ripple").inE().has("weight", gte(0.5)).outV().properties()
+----
+
+The next code block shows the same queries, but with specified keys and labels.
+
+[gremlin-groovy,existing]
+----
+g.V().has("person","name","marko").out("created","knows")
+g.V().has("person","name","marko").out("created").valueMap("name","lang")
+g.V().has("software","name","ripple").inE("created").has("weight", gte(0.5)).outV().
+  properties("name","age")
+----
+
+[[unnecessary-steps]]
+== Unnecessary Steps
+
+There are quite a few steps and patterns that can be combined into a much shorter form. TinkerPop is trying to optimize queries, by
+rewriting such patterns automatically using traversal optimization strategies. These strategies, however, do have a few preconditions
+and under certain circumstance they will not attempt to rewrite a traversal. For example, if the traversal has path computations
+enabled (e.g. by using certain steps, such as `path()`, `simplePath()`, `otherV()`, etc.), then the assumption is that all steps are
+required in order to produce the desired path.
+
+An often seen anti-pattern is the one that explicitely traversers to an edge and then to a vertex without using any filters.
+
+[gremlin-groovy,modern]
+----
+g.V().hasLabel("person").outE("created").inV().dedup()    <1>
+g.V().hasLabel("software").inE("created").outV().count()  <2>
+----
+
+<1> The `created` edge is never really needed as the traversal only asks for all things that were created by all persons in the graph.
+    These "things" are only represented by the adjacent vertices, not the edges.
+<2> This traversals counts the persons in the graph who created software. The interesting thing about this query is that it actually 
+    doesn't need to traverse all the way to the `person` vertices to count them. In this case it's sufficient to count the edges
+    between the `software` and `person` vertices. The performance of this query pretty much depends on the particular provider
+    implementation, but counting incident edges is usually much faster than counting adjacent vertices.
+
+The next code block shows the two aforementioned queries properly rewritten.
+
+[gremlin-groovy,modern]
+----
+g.V().hasLabel("person").out("created").dedup()
+g.V().hasLabel("software").inE("created").count()
+----
+
+Another anti-pattern that is commonly seen is the chaining of `where()`-steps using predicates. Consider the following traversal:
+
+[gremlin-groovy,modern]
+----
+g.V().as('a').
+  both().where(lt('a')).by(id).as('b').
+  both().where(lt('a')).by(id).where(gt('b')).by(id).as('c').
+  not(both().where(eq('a'))).
+  select('a','b','c').
+    by('name')
+----
+
+Ignoring the anti-patterns that were discussed before there's not much wrong with the traversal, but note the two chained `where()`-steps
```

Comment:
I feel like there's maybe two different things that you mean here so I can't say what the correction is but there's something awkward about how this phrase is written:

> Ignoring the anti-patterns that were discussed before there's not much wrong with the traversal,

Reply:
"Ignoring the anti-patterns that were discussed before" -- I added this because the queries use `both()` (instead of `both(<label>)`), which was declared to be an anti-pattern in an earlier section. I can change it to `both("knows","created")`, but I wanted the reader to focus on the `where()` steps, hence I kept the other steps as short as possible.

Reply:
oh...now i get it. i think i was pausing my reading in the wrong place. i think a comma after "before" would help:

"Ignoring the anti-patterns that were discussed before, there's not much wrong with the traversal, but note the two chained `where()`-steps..."